src/language/patterns.md

```diff
@@ -0,0 +1,358 @@
+
+## Uses cases for patterns
+
+The [section above](#places-patterns-can-appear)
+describes _how_ you can use patterns in Dart code constructs, 
+sometimes with interesting uses cases as examples (like [swapping](#variable-assignment)
+the values of two variables, or
+[destructuring key-value pairs](#for-and-for-in-loops)
+in a map). 
+
+This section describes even more use cases, answering _when and why_ you might
+want to use patterns, the kinds of problems they solve, and the idioms they're 
+best suited for.
+
+### Destructuring multiple returns
+
+[Records][] allow aggregating and returning multiple values from a single function
+call. Patterns add the ability to destructure a record's fields
+directly into local variables, inline with the function call.
+
+Instead of individually declaring new local variables for each record field,
+like this:
+
+```dart
+main() {
+  var info = userInfo(json);
+  var name = info.$1;
+  var age = info.$2;
+}
+```
+
+You can destructure the fields of a record that a function returns into local
+variables using a [variable declaration](#variable-declaration) or
+[assigment pattern](#variable-assignment), and a record pattern as its subpattern:
+
+```dart
+main() {
+  var (name, age) = userInfo(json);
+}
+```
+
+### Destructuring class instances
+
+[Object patterns][object] match against named object types, allowing
+you to destructure their data using the getters the object's class already exposes.
+
+To destructure an instance of a class, use the named type, followed by the properties
+to destructure enclosed in parenthesis:
+
+```dart
+main() {
+  final Foo myFoo = Foo(one: 'one', two: 2);
+  var Foo(:one, :two) = myFoo;
+  print('one $one, two $two');
+}
+```
+
+### Algebraic datatypes 
+
+Object destructuring and switch cases are conducive to writing
+code in an [algebraic datatype][] style: 
+when you have a family of related types and an operation that needs
+specific behavior for each type.
+
+Instead of implementing the operation as an instance method for every respective type,
+you can keep the operations in a single function, switching over subtypes for
+variations in behavior:
+
+```dart
+sealed class Shape {}
+
+class Square implements Shape {
+  final double length;
+  Square(this.length);
+}
+
+class Circle implements Shape {
+  final double radius;
+  Circle(this.radius);
+}
+
+double calculateArea(Shape shape) =>
+  switch (shape) {
+    Square(length: var l) => l * l,
+    Circle(radius: var r) => math.pi * r * r
+  };
+```
+
+### Validating incoming JSON
+
+[Map][] and [list][] patterns work well for destructuring key-value pairs in
+JSON data:
+
+```dart 
+var json = {'user': ['Lily', 13]};
+var {'user': [name, age]} = json;
+```
+
+The example above makes sense when you already know that the JSON
+data has the structure you expect.
+If the data comes from an external source, like over the network as is commonly
+the case, you can't be sure of it's structure and
+should validate it first. 
```

Comment:
**issues:** Write as shorter sentences per [GSG](https://developers.google.com/style/accessibility#ease-of-reading).
**typo:** `its` not `it's`.
```suggestion
If the data comes from an external source, you can't validate its structure.
You should validate it first. 
```